crates/consensus/beacon/src/engine/mod.rs

```diff
@@ -255,8 +255,51 @@ where
 
         // TODO: check PoW / EIP-3675 terminal block conditions for the fork choice head
         // TODO: ensure validity of the payload (is this satisfied already?)
-
         let is_first_forkchoice = self.forkchoice_state.is_none();
+
+        // Client software MUST return -38002: Invalid forkchoice state error if the payload
+        // referenced by forkchoiceState.headBlockHash is VALID and a payload referenced by either
+        // forkchoiceState.finalizedBlockHash or forkchoiceState.safeBlockHash does not belong to
+        // the chain defined by forkchoiceState.headBlockHash.
+
+        // TODO: is this the right check that the block exists and is VALID?
+        // This is here so we know that the block is `VALID` before performing same-chain checks,
+        // if it's in the tree or db it should be VALID
+        if !self.blockchain_tree.is_block_known(&state.head_block_hash)? {
+            // If this is the first forkchoice received, start downloading from safe block
+            // hash.
+            let target = if is_first_forkchoice &&
+                !state.safe_block_hash.is_zero() &&
+                self.get_block_number(state.safe_block_hash)?.is_none()
+            {
+                PipelineTarget::Safe
+            } else {
+                PipelineTarget::Head
+            };
+            self.forkchoice_state = Some(state);
+            // TODO: does the pipeline need to be idle here?
+            self.require_pipeline_run(target);
+            return Ok(OnForkChoiceUpdated::valid(PayloadStatus::from_status(
+                PayloadStatusEnum::Syncing,
+            )))
+        }
+
+        // check if finalized is part of the same chain (and not zero)
+        if !state.finalized_block_hash.is_zero() &&
+            !self
+                .blockchain_tree
+                .share_chain(&state.head_block_hash, &state.finalized_block_hash)?
+        {
+            return Ok(OnForkChoiceUpdated::invalid_state())
```

Comment:
is this right that we can trigger the pipeline and return an invalid state?

Reply:
I'm not sure, in this case we may have `safe`, `finalized`, and `head` hashes that correspond to valid blocks in their own separate sidechains, and we may have to trigger the pipeline to retrieve or execute those blocks. When sent together as a single forkchoice update, the update is invalid if the `safe` and `head`, or `finalized` and `head`, do not belong to the same sidechain. So if we have to "set" the forkchoice state to run the pipeline, that forkchoice state could be invalid / inconsistent